flat_mutation_reader.cc

```diff
@@ -955,6 +956,105 @@ make_flat_mutation_reader_from_fragments(schema_ptr schema, reader_permit permit
+        flat_mutation_reader _rd;
+        const dht::partition_range* _pr;
+        const query::partition_slice* _slice;
+        dht::ring_position_comparator _cmp;
+        std::optional<clustering_ranges_walker> _ranges_walker;
+
+        future<> _started = make_ready_future<>();
+
+    public:
+        reader(flat_mutation_reader rd, const dht::partition_range& pr, const query::partition_slice& slice)
+            : flat_mutation_reader::impl(rd.schema(), rd.permit())
+            , _rd(std::move(rd))
+            , _pr(&pr)
+            , _slice(&slice)
+            , _cmp(*_schema) {
+        }
+
+        virtual future<> fill_buffer(db::timeout_clock::time_point timeout) override {
+            shared_future<> f = do_until([this] { return is_buffer_full() || is_end_of_stream(); }, [this, timeout] {
+                return _rd.fill_buffer(timeout).then([this] {
+                    auto f = make_ready_future<>();
+                    
+                    while (!_rd.is_buffer_empty()) {
+                        auto mf = _rd.pop_mutation_fragment();
+                        switch (mf.mutation_fragment_kind()) {
+                        case mutation_fragment::kind::partition_start: {
+                            auto& dk = mf.as_partition_start().key();
+                            if (!_pr->contains(dk, _cmp)) {
+                                f = _rd.next_partition();
+                                break;
+                            } else {
+                                _ranges_walker.emplace(*_schema, _slice->row_ranges(*_schema, dk.key()), false);
+                            }
+                            // fall-through
+                        }
+
+                        case mutation_fragment::kind::static_row: // fall-through
+                        case mutation_fragment::kind::partition_end: // fall-through
+                            push_mutation_fragment(std::move(mf));
+                            break;
+
+                        case mutation_fragment::kind::clustering_row:
+                            if (_ranges_walker->advance_to(mf.position())) {
+                                push_mutation_fragment(std::move(mf));
+                            }
+                            break;
+
+                        case mutation_fragment::kind::range_tombstone:
+                            auto&& rt = mf.as_range_tombstone();
+                            if (_ranges_walker->advance_to(rt.position(), rt.end_position())) {
+                                push_mutation_fragment(std::move(mf));
+                            }
+                            break;
+                        }
+                    }
+                    
+                    return f.then([this] {
+                        _end_of_stream = _rd.is_end_of_stream();
+                    });
+                });
+            });
+
+            _started = _started.then([ft = std::move(f.get_future())] () mutable { return std::move(ft); });
```

Comment:
I don't understand why this is needed.

Reply:
my bad, I forgot to return it inside the `close()` method
We use `_started` to make sure all the `fill_buffer` operations can be waited on before destroying the reader.

Reply:
ah, if so then perfect, let's just close the underlying reader in that method then